### PR TITLE
Graphical editor bugfixes

### PR DESCRIPTION
**Pull request by Daedo**
Multiselect Mega Monster Merge 2 - The Reckonning:
Redo Merge https://github.com/junkerm/specmate/commit/82eedb62f2fd61fe0ed3a176d574af7e78d00f86
Bugfixes

### CODE REVIEW

File: web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        let chain = Promise.resolve();
+        for (const element of elements) {
+            chain = chain.then(() => this.deleteElement(element, compoundId));
+        }
+        chain.then(() => this.selectionService.deselectElements(elements));
+        return chain;
+    }
+
+    private deleteElement(element: IContainer, compoundId: string): Promise<void> {
+        this.dataService.readElement(Url.parent(element.url), true).then((model: IContainer) => this.selectionService.select(model));
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts]: is the readElement necessary? Can you add a comment why?

File: web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        return Promise.resolve();
+    }
+
+    public deleteElementAndDeselect(element: IContainer, compoundId: string): Promise<void> {
+        this.selectionService.deselectElement(element);
+        return this.deleteElement(element, compoundId);
+    }
+
+    private deleteNode(node: IModelNode, compoundId: string): Promise<void> {
+        return this.dataService.readContents(this.parent.url, true).then( (content: IContainer[]) => {
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts]: as above: readElement necessary? please add comment why

File: web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    private deleteConnection(connection: IModelConnection, compoundId: string): Promise<void> {
+        return this.removeConnectionFromSource(connection, compoundId)
+            .then(() => this.removeConnectionFromTarget(connection, compoundId))
+            .then(() => this.dataService.deleteElement(connection.url, true, compoundId));
+    }
+
+    private removeConnectionFromSource(connection: IModelConnection, compoundId: string): Promise<void> {
+        return this.dataService.readElement(connection.source.url, true)
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts]: see above

File: web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        return this.dataService.readElement(connection.target.url, true)
+            .then((target: IModelNode) => {
+                let proxyToDelete: Proxy = target.incomingConnections.find((proxy: Proxy) => proxy.url === connection.url);
+                Arrays.remove(target.incomingConnections, proxyToDelete);
+                return target;
+            })
+            .then((target: IContainer) => this.dataService.updateElement(target, true, compoundId));
+    }
+
+    // Create
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts]: Please add a sensible comment

File: web/src/app/modules/views/side/modules/selected-element/services/selected-element.service.ts [hunk 5]
```diff
@@ -20,36 +22,89 @@ export class SelectedElementService {
...
         if (!this._selectionChanged) {
-            this._selectionChanged = new EventEmitter<IContainer>();
+            this._selectionChanged = new EventEmitter<IContainer[]>();
         }
         return this._selectionChanged;
     }
 
     public get hasSelection(): boolean {
-        return this.selectedElement !== undefined;
+        return this._selectedElements.length > 0;
```
Comment (junkerm) [inline web/src/app/modules/views/side/modules/selected-element/services/selected-element.service.ts]: can _selectedelements still be undefined?

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/common/generic-graphical-connection.component.svg [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<ng-container [ngSwitch]="connection.className">
+    <svg:g *ngSwitchCase="'CEGConnection'" ceg-graphical-connection [connection]="connection" [nodes]="nodes"></svg:g>
+    
+    <svg:g *ngSwitchCase="'ProcessConnection'" process-graphical-connection [connection]="connection" [nodes]="nodes"></svg:g>
+    
+    <svg:g *ngSwitchDefault>
```
[the remark this answers is not in the document]
Comment (Daedo) [reply web/src/app/modules/views/main/editors/modules/graphical-editor/components/common/generic-graphical-connection.component.svg]: I think having a default case that indicates that something went wrong is a lot easier to debug than not drawing anything.

[the remark this answers is not in the document]
Comment (junkerm) [reply web/src/app/modules/views/main/editors/modules/graphical-editor/components/common/generic-graphical-connection.component.svg]: ok

**Comment by junkerm** [comment]
Shouldnt element and model be the same thing? Am 08.10.2018 09:26 schrieb Daedo <notifications@github.com>:@Daedo commented on this pull request.



In web/src/app/modules/views/main/editors/modules/tool-pallette/util/graphTransformer.ts:
> +    constructor(private dataService: SpecmateDataService, private selectionService: SelectedElementService, private parent: IContainer) {
+        this.elementProvider = new ElementProvider(this.parent);
+    }
+
+    // Delete
+    public deleteAll(elements: IContainer[], compoundId: string): Promise<void> {
+        let chain = Promise.resolve();
+        for (const element of elements) {
+            chain = chain.then(() => this.deleteElement(element, compoundId));
+        }
+        chain.then(() => this.selectionService.deselectElements(elements));
+        return chain;
+    }
+
+    private deleteElement(element: IContainer, compoundId: string): Promise<void> {
+        this.dataService.readElement(Url.parent(element.url), true).then((model: IContainer) => this.selectionService.select(model));

I think so. I don't think there is another easy way to resolve the model url to get the IContainer.

File: web/src/app/modules/views/side/modules/selected-element/services/selected-element.service.ts (hunk 5 above)
Comment (Daedo) [reply web/src/app/modules/views/side/modules/selected-element/services/selected-element.service.ts]: No it is initialized as an empty list and the setter doesn't allow it to be set to undefined.

**Comment by Daedo** [comment]
No. 
We use Url.parent()
So we have an element of the model. Like a node or a connection and from that we derive the model the element is contained in. The element is not the same thing and we can’t let the selection as it is since we’re deleting the element.

**Pull request merged**